src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +407,60 @@ where
+                                                // dummy data
+                                                row_packer.extend(
+                                                    (0..key_arity).map(|_| Datum::Dummy),
+                                                );*/
+                                                row_packer.extend(full_row.iter().skip(key_arity));
+                                                row_packer.finish_and_reuse()
+                                            });
+                                            (
+                                                Some(thinned_value.clone()),
+                                                current_values.insert(
+                                                    decoded_key.clone(),
+                                                    // Thin out the row to not contain a copy of the key(s)
+                                                    thinned_value,
+                                                ),
+                                            )
+                                        } else {
+                                            (None, current_values.remove(&decoded_key))
+                                        };
+
+                                    // This closure re-uses a `Row` to
+                                    // rebuild the full `Row` with both keys and values, before we
+                                    // send them off to sql-land
+                                    let mut repack_value = |row: Row| {
+                                        row_packer.clear();
+
+                                        row_packer.extend(decoded_key.iter());
+                                        row_packer.extend(row.iter());
+
+                                        row_packer.finish_and_reuse()
                                     };
-                                    if let Some(old_value) = old_value {
+
+                                    if let Some(thinned_old_value) = thinned_old_value {
                                         // retract old value
-                                        session.give((old_value, cap.time().clone(), -1));
+                                        session.give((
+                                            thinned_old_value.map(&mut repack_value),
+                                            cap.time().clone(),
+                                            -1,
+                                        ));
                                     }
-                                    if let Some(new_value) = new_value {
+                                    if let Some(thinned_new_value) = thinned_new_value {
                                         // give new value
-                                        session.give((new_value, cap.time().clone(), 1));
+                                        session.give((
+                                            thinned_new_value.map(&mut repack_value),
```

Comment:
Can't you just use the original `new_value` here, rather than recovering it from `thinned_new_value` ?